enterprise/frontend/src/metabase-enterprise/metabot/state/reducer.ts

```diff
@@ -0,0 +1,29 @@
+import { type PayloadAction, createSlice } from "@reduxjs/toolkit";
+
+import type { ChatMessage, MetabotAgentChatContext } from "metabase-types/api";
+
+export interface MetabotState {
+  chat: {
+    history: ChatMessage[];
+    context: MetabotAgentChatContext;
```

Comment:
I think that `context` is not a state property; it should be a selector

Reply:
`type Message = string | { message: string, context: Ctx | (store: Store) => Ctx }`
`sendMessage(msg: Message)`
something like this?

Reply:
context would default to an empty object if not provided

Reply:
something like:
```
type MetabotContextProvider = (state: State) => MetabotAgentChatContext;

const PLUGIN_METABOT = {
  ...
  contextProviders: MetabotContextProvider[];
}

function getMetabotContext(state: State): MetabotAgentChatContext {
  return PLUGIN_METABOT.contextProviders.reduce((provider, context) => ({ ...context, provider(state)}), {});
}
```

Reply:
it's a bit unclear how providers should be registered - and where. There could be a list of functions in EE code that extracts stuff from redux; or we can even make top-level app components do that